### PR TITLE
Bean Import support for JSR 330

### PR DESCRIPTION
**Pull request by graemerocher**
Fixes https://github.com/micronaut-projects/micronaut-core/issues/1885

### CODE REVIEW

File: graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java [hunk 1]
```diff
@@ -91,15 +99,36 @@ public int getOrder() {
...
+                originatingElements.add(element);
                 packages.add(element.getPackageName());
                 final String beanName = element.getName();
                 addBean(beanName);
-                resolveClassData(beanName + "[]");
+                if (element.isArray()) {
+                    resolveClassData(beanName + "[]");
+                } else {
+                    resolveClassData(beanName);
+                }
```
Comment (graemerocher) [inline graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java]: @ilopmar this appeared to be a bug? Please confirm

File: graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java [hunk 2]
```diff
@@ -160,8 +217,9 @@ public void visitMethod(MethodElement element, VisitorContext context) {
     @Override
     public void visitConstructor(ConstructorElement element, VisitorContext context) {
         if (!isSubclass) {
-            if (element.hasAnnotation(Creator.class)) {
+            if (element.hasAnnotation(Creator.class) && element.isPrivate()) {
```
Comment (graemerocher) [inline graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java:215]: @ilopmar again I am not sure why we were creating reflective data is the creator didn't require reflection

File: graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java (hunk 1 above)
Comment (ilopmar) [reply graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java]: Someone submitted a PR sometime ago to also include arrays in the reflection config for all the declared classes. Apparently they needed for their use case

File: graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java (hunk 2 above)
Comment (ilopmar) [reply graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java:215]: I don't remember, probably it was necessary sometime ago and maybe it's not now.

File: graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java [hunk 3]
```diff
@@ -122,11 +122,7 @@ public void visitClass(ClassElement element, VisitorContext context) {
                 packages.add(element.getPackageName());
                 final String beanName = element.getName();
                 addBean(beanName);
-                if (element.isArray()) {
-                    resolveClassData(beanName + "[]");
-                } else {
-                    resolveClassData(beanName);
-                }
+                resolveClassData(beanName + "[]");
```
Comment (ilopmar) [inline graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java]: @graemerocher shouldn't this be the other way? I mean remove the one with the array and keep the original?

**Comment by graemerocher** [comment]
> shouldn't this be the other way? I mean remove the one with the array and keep the original?

I just reverted to the original behaviour

File: inject/src/main/java/io/micronaut/context/annotation/Import.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+     * A list of package names to import.
+     *
+     * <p>Note that only types with a bean defining annotation will be imported.</p>
+     *
+     * @return The packages to import.
+     */
+    String[] packages() default {};
+
+    /**
+     * The annotations types to include in a search. If set to "*" will include all. Defaults to only included types annotated with scopes or qualifiers.
```
Comment (jameskleeh) [inline inject/src/main/java/io/micronaut/context/annotation/Import.java]: Perhaps make this more clear that this only works with packages also being set

**Pull request merged**